Implement maximum length on the multiline text control under GTK

oneeyeman1 opened the pull request:
This is a second attempt.

It is simpler and less code.

Please review

oneeyeman1 committed:
Initial implementation

oneeyeman1 committed:
Whitespaces fix

oneeyeman1 commented:
@vadz ,
"Not enough compute credits to prioritize tasks!"

I don't know what it means, but it sounds serious...

Thx.

lanurmi commented on src/gtk/textctrl.cpp:
```diff
@@ -684,7 +706,6 @@ void wxTextCtrl::Init()
 {
     m_dontMarkDirty =
     m_modified = false;
-
```
Meaningless whitespace change

lanurmi commented on src/gtk/textctrl.cpp:
```diff
@@ -28,7 +28,7 @@
 #include <sys/stat.h>
 #include <ctype.h>
-
+#include <string.h>
```
A (probably intentional) empty line removed

lanurmi commented on src/gtk/textctrl.cpp:
```diff
@@ -873,7 +897,7 @@ bool wxTextCtrl::Create( wxWindow *parent,
                               G_CALLBACK (au_apply_tag_callback), nullptr);
 
-            // Check for URLs in the initial string passed to Create
+            // Check for URLs in the initial string sed to Create
```
Why was this included in the commit?

lanurmi commented on src/gtk/textctrl.cpp:
```diff
@@ -1252,10 +1289,11 @@ void wxTextCtrl::WriteText( const wxString &text )
-
-    const wxScopedCharBuffer buffer(text.utf8_str());
+    auto temp = text;
+    const wxScopedCharBuffer buffer(temp.utf8_str());
```
Why introduce a new variable for no reason?

lanurmi commented on src/gtk/textctrl.cpp:
```diff
@@ -1766,7 +1805,10 @@ void wxTextCtrl::Paste()
     if ( IsMultiLine() )
+    {
         g_signal_emit_by_name (m_text, "paste-clipboard");
+        auto value = GetValue();
```
`value` is assigned, but never read. This piece doesn't appear to do anything in practice.

oneeyeman1 committed:
Implement code review suggestions

oneeyeman1 commented:
@lanurmi ,
Thank you for the review.
I addressed all this in the latest commit.

oneeyeman1 committed:
Fix whitespaces

oneeyeman1 committed:
Partially restore code from the original PR

MaartenBent commented on include/wx/gtk/textctrl.h:
```diff
@@ -59,7 +59,7 @@ class WXDLLIMPEXP_CORE wxTextCtrl: public wxTextCtrlBase
     virtual void GetSelection(long* from, long* to) const override;
 
     virtual void Remove(long from, long to) override;
-
```
you removed an empty line

MaartenBent commented on include/wx/gtk/textctrl.h:
```diff
@@ -233,7 +234,7 @@ class WXDLLIMPEXP_CORE wxTextCtrl: public wxTextCtrlBase
     // Our text buffer. Convenient, and holds the buffer while using
     // a dummy one when frozen
     GtkTextBuffer *m_buffer;
-
```
you removed an empty  line

MaartenBent commented on include/wx/gtk/textentry.h:
```diff
@@ -62,7 +62,7 @@ class WXDLLIMPEXP_CORE wxTextEntry : public wxTextEntryBase
     void SendMaxLenEvent();
     bool GTKEntryOnInsertText(const char* text);
     bool GTKIsUpperCase() const { return m_isUpperCase; }
-
```
you removed an empty line

MaartenBent commented on include/wx/gtk/textentry.h:
```diff
@@ -99,7 +99,7 @@ class WXDLLIMPEXP_CORE wxTextEntry : public wxTextEntryBase
     // Call this from the overridden wxWindow::GTKIMFilterKeypress() to use
     // GtkEntry IM context.
     int GTKEntryIMFilterKeypress(GdkEventKey* event) const;
-
```
you removed an empty line

MaartenBent commented on samples/text/text.cpp:
```diff
@@ -140,7 +140,7 @@ class MyPanel: public wxPanel
     MyTextCtrl    *m_tab;
     MyTextCtrl    *m_readonly;
     MyTextCtrl    *m_limited;
-
```
you removed an empty line

MaartenBent commented on samples/text/text.cpp:
```diff
@@ -1178,6 +1178,9 @@ MyPanel::MyPanel( wxFrame *frame, int x, int y, int w, int h )
     m_limited->SetMaxLength(8);
     wxSize size2 = m_limited->GetSizeFromTextSize(m_limited->GetTextExtent("WWWWWWWW"));
     m_limited->SetSizeHints(size2, size2);
+    m_limitedMultiline = new MyTextCtrl( this, wxID_ANY, "", wxPoint( 10, 110 ), wxDefaultSize, wxTE_MULTILINE );
```
```suggestion
    m_limitedMultiline = new MyTextCtrl(this, wxID_ANY, "", wxPoint(10, 110), wxDefaultSize, wxTE_MULTILINE);
```

No need for the extra spaces in arguments, see [coding guidelines](https://www.wxwidgets.org/develop/coding-guidelines/#spaces_keywords):
`Do not use spaces in these positions for the function calls but still use them to separate function arguments`
 This happens throughout the PR, so I'll only comment it here.

MaartenBent commented on src/gtk/textctrl.cpp:
```diff
@@ -29,6 +29,7 @@
 #include <sys/stat.h>
 #include <ctype.h>
 
+#include <string.h>
```
Is this needed? I don't see any usage of std::string.

MaartenBent commented on src/gtk/textctrl.cpp:
```diff
@@ -560,6 +561,28 @@ au_insert_text_callback(GtkTextBuffer *buffer,
+        // reason, ("Run Last") it won't work in GTKOnInsertText() as it called from
+        // the handler that does not connected after
+        gtk_text_buffer_delete( buffer, &offset, end );
+#if GTK_CHECK_VERSION( 3, 0, 0 )
```
Only since 3.2? https://docs.gtk.org/gtk3/method.TextIter.assign.html
Also needs runtime checks?

MaartenBent commented on src/gtk/textctrl.cpp:
```diff
@@ -1226,12 +1263,14 @@ static gboolean afterLayout(void* data)
 void wxTextCtrl::WriteText( const wxString &text )
 {
     wxCHECK_RET( m_text != nullptr, wxT("invalid text ctrl") );
-
```
you removed an empty line

MaartenBent commented on src/gtk/textctrl.cpp:
```diff
@@ -1764,8 +1803,7 @@ void wxTextCtrl::Copy()
 void wxTextCtrl::Paste()
 {
     wxCHECK_RET( m_text != nullptr, wxT("invalid text ctrl") );
-
```
you removed an empty line

MaartenBent commented on src/gtk/textctrl.cpp:
```diff
@@ -1226,12 +1263,14 @@ static gboolean afterLayout(void* data)
 {
     wxCHECK_RET( m_text != nullptr, wxT("invalid text ctrl") );
-
+    auto old = m_maxLengthAllowed;
```
Use a more descriptive variable name

oneeyeman1 replied:
@MaartenBent ,
You have something in mind?

MaartenBent replied:
Something like `oldMaxLengthAllowed`. So it describes what the variables is...

MaartenBent commented on src/gtk/textctrl.cpp:
```diff
@@ -1252,11 +1291,10 @@ void wxTextCtrl::WriteText( const wxString &text )
         return;
     }
 
-    const wxScopedCharBuffer buffer(text.utf8_str());
```
what is the reason you got rid of the `utf8_str()` and use `text` directly?

MaartenBent commented on src/gtk/textctrl.cpp:
```diff
@@ -2153,13 +2191,24 @@ wxSize wxTextCtrl::DoGetSizeFromTextSize(int xlen, int ylen) const
     {
+        // add space for vertical scrollbar
+        if ( m_scrollBar[1] && !(m_windowStyle & wxTE_NO_VSCROLL) )
+            tsize.IncBy(GTKGetPreferredSize(GTK_WIDGET(m_scrollBar[1])).x + 3, 0);
```
what is this `3`? Same for horizontal scrollbar.

oneeyeman1 replied:
@MaartenBent ,
I have no idea. This is an original code.

MaartenBent replied:
What do you mean? You added it here. Why?

MaartenBent commented on src/gtk/textctrl.cpp:
```diff
@@ -2153,13 +2191,24 @@ wxSize wxTextCtrl::DoGetSizeFromTextSize(int xlen, int ylen) const
+        if ( !HasFlag(wxBORDER_NONE) )
+        {
+            // hardcode borders, margins, etc
+            tsize.IncBy(5, 4);
```
Where do these numbers come from?

oneeyeman1 replied:
@MaartenBent ,
Same as with previous comment

MaartenBent replied:
It is new code, added by you. So please explain why these values.

MaartenBent commented on src/gtk/textctrl.cpp:
```diff
@@ -1764,8 +1803,7 @@ void wxTextCtrl::Copy()
     wxCHECK_RET( m_text != nullptr, wxT("invalid text ctrl") );
-
-    if ( IsMultiLine() )
+    if( IsMultiLine() )
```
this should have a space after `if`, see coding guidelines.

oneeyeman1 commented:
@MaartenBent ,
Thx for the review.
Everything will be fixed in the upcoming commit.

Except I changed the variable name but probably its still not good. And I have no idea about the numbers.

oneeyeman1 committed:
Fix code review suggestions

MaartenBent commented:
Did you read my reply in the mailing list about why the tests randomly fail? Can you include that fix in this PR?

For reference:
> Some more details, character 100 in linePattern variable in the test case is not initialized.
It fills from 0 to 99 with values, and then sets 101 to \0.

> I'd just simplify it to something like:
```c++
wxString linePattern;
for (i = 0; i < lenPattern; i++)
{
    linePattern.append(wxChar('0' + i % 10));
}
```

oneeyeman1 commented:
@MaartenBent ,
> Did you read my reply in the mailing list about why the tests randomly fail? Can you include that fix in this PR?
> 
> For reference:
> 
> > Some more details, character 100 in linePattern variable in the test case is not initialized.
> > It fills from 0 to 99 with values, and then sets 101 to \0.
> 
> > I'd just simplify it to something like:
> 
> ```c++
> wxString linePattern;
> for (i = 0; i < lenPattern; i++)
> {
>     linePattern.append(wxChar('0' + i % 10));
> }
> ```

Yes I did.
That code with utf8_str() is restored and it seems to be passing now.
Changing the code in test suite - why? It is unrelated.
When this is merged you can make it like this.
Besides, it might be intentional for whatever reasons.

MaartenBent commented:
Because it is testing with a string with uninitialized data. But whatever, I can fix it.
I doubt @vadz added this intentionally in e7f2eb8f830d90cea7c77ce937888c1673356ca4

MaartenBent commented:
> Also - kind of unrelated - how did you access the test run logs?

I just click on the CI details
<img width="366" alt="step1" src="https://github.com/user-attachments/assets/24c93b27-cb73-49dd-889a-3059156f7a2e">


Then it shows the full log with the errors
<img width="924" alt="step2" src="https://github.com/user-attachments/assets/13833d14-1265-44be-9c7b-f5e662394b36">

And you can also see the weird character at the end of the test line, that is caused by the uninitialized value
<img width="534" alt="step3" src="https://github.com/user-attachments/assets/cf3184b9-bfb1-40cb-85c9-49be14b02ed1">

oneeyeman1 commented:
@MaartenBent ,
Apparently you were right. It was added by me.
Completely unintentional though.

I removed both and it still seems to work.

I'll wait for the build to finish and push it.

Thank you and sorry for not checking before replying.

In regards to logs - I thought you were talking about the actual on the hard drive as the tests are run with everything redirected to file.

But that's OK - the point is mute now.

Thx once again.

oneeyeman1 commented:
It also would be interesting to know why some builds are queued... ;-)

oneeyeman1 committed:
Rename the variable as approriate and remove accidental additions

oneeyeman1 commented:
Lauri and Maarten,
Thank you for the review.

Hopefully it can be applied soon.

oneeyeman1 commented:
@MaartenBent ,
> Did you read my reply in the mailing list about why the tests randomly fail? Can you include that fix in this PR?
> 
> For reference:
> 
> > Some more details, character 100 in linePattern variable in the test case is not initialized.
> > It fills from 0 to 99 with values, and then sets 101 to \0.
> 
> > I'd just simplify it to something like:
> 
> ```c++
> wxString linePattern;
> for (i = 0; i < lenPattern; i++)
> {
>     linePattern.append(wxChar('0' + i % 10));
> }
> ```

You can make your PR on top.
Or just wait for this to be applied...

MaartenBent commented:
When I check the changed files, there are still modifications to `DoGetSizeFromTextSize`. Did you intent to revert them all?

oneeyeman1 committed:
Restore accidentally chnged function

oneeyeman1 commented:
@MaartenBent ,
Should now be fixed. Sorry about the breakage.


Can you do me a favor and try this PR? See if it works for you?

Thank you.

MaartenBent commented:
Seems to work in the text sample. But there is a difference with the single line limited text control. 
The single line only has event:
```
You can't enter more characters into this control.
```
But the multi line has two events:
```
You can't enter more characters into this control.
Text changed in some control
```
On Windows both have only the first event.

Also when pasting a too big text, the single line doesn't fill any text. The multi line fills until the limit. On Windows both fill to the limit, so I guess the single line is incorrect. So maybe not relevant for this PR.

Also maybe you wan't to handle `m_limitedMultiline` in `OnSetEnabled` same as `m_limited`.

oneeyeman1 commented:
@MaartenBent ,
Thx for testing.

> Seems to work in the text sample. But there is a difference with the single line limited text control. The single line only has event:
> 
> ```
> You can't enter more characters into this control.
> ```
> 
> But the multi line has two events:
> 
> ```
> You can't enter more characters into this control.
> Text changed in some control
> ```
> 
> On Windows both have only the first event.

I'll see what can be done.

I have an idea how to fix it - we will see.

> Also when pasting a too big text, the single line doesn't fill any text. The multi line fills until the limit. On Windows both fill to 
> the limit, so I guess the single line is incorrect. So maybe not relevant for this PR.
>
> Also maybe you wan't to handle m_limitedMultiline in OnSetEnabled same as m_limited.

It is not. That's native control behavior.
However, I'll probably add the note in the docs about this differences.

oneeyeman1 committed:
Fix event generation if the control reaches maximum length

oneeyeman1 commented:
@MaartenBent ,
Event generation should now be fixed.

I forgot to update the docs in regards to the single line text control,

Can you update the code and re-test.

Thank you,

MaartenBent commented on include/wx/gtk/textentry.h:
```diff
@@ -100,6 +101,7 @@ class WXDLLIMPEXP_CORE wxTextEntry : public wxTextEntryBase
     // GtkEntry IM context.
     int GTKEntryIMFilterKeypress(GdkEventKey* event) const;
 
+    int m_maxlen = 0;
```
Is there a reason this is `protected` instead of `private`?
If so, maybe separate it from the unrelated comments below. If not, move it to the other variable declarations.

oneeyeman1 replied:
@MaartenBent ,
I don't remember.

I think it should be private.

Will fix.

oneeyeman1 replied:
@MaartenBent ,
It is declared protected because SetMaxLength is a function inside wxTextCtrl to set it.

For now I will keep it protected. I might change it when I go to Mac implementation and put it in the main header...

MaartenBent replied:
I think generally you make something private unless there is a reason that it should be protected or public.

MaartenBent commented on src/gtk/textctrl.cpp:
```diff
@@ -2160,7 +2206,7 @@ wxSize wxTextCtrl::DoGetSizeFromTextSize(int xlen, int ylen) const
         gtk_widget_get_preferred_size(m_widget, &req, nullptr);
         tsize.IncTo(wxSize(req.width, req.height));
-    }
+}
```
wrong indentation

MaartenBent commented on include/wx/gtk/textctrl.h:
```diff
@@ -234,6 +236,7 @@ class WXDLLIMPEXP_CORE wxTextCtrl: public wxTextCtrlBase
     // a dummy one when frozen
     GtkTextBuffer *m_buffer;
 
+    bool m_maxLengthAllowed = false;
```
why is this extra variable needed? Can't you just check if max length > 0? I removed all the `IsMaxLengthAllowed` and `m_maxLengthAllowed` and it still seems to work fine.

oneeyeman1 replied:
@MaartenBent ,
All 3 major events - typing on the keyboard, pasting from the clipboard and setting/appending text programmatically goes thru this routine.

Now, while first 2 are fine - third one is not.
Try to run the test without this - it will fail.

The reason for that is because MS, for whatever reason, decided to check the max length for multiline text control only on the user action.
So in order for the test to pass I need to add the extra variable.

Just try to compare the test run on MS and GTK without this variable.

MaartenBent commented on src/gtk/textctrl.cpp:
```diff
@@ -797,8 +820,11 @@ bool wxTextCtrl::Create( wxWindow *parent,
         // new, empty control, see https://github.com/wxWidgets/wxWidgets/issues/11409
         gtk_entry_get_text((GtkEntry*)m_text);
 
+#ifndef __WXGTK3__
```
Why is this added? Limiting border related stuff seems unrelated to max length.

oneeyeman1 replied:
@MaartenBent ,
Will check.

MaartenBent reviewed (changes requested):
Events seem fine now. But I added some more comments.

Also can you mark the fixed comments as resolved, so the PR is easier to review.

MaartenBent commented on include/wx/gtk/textentry.h:
```diff
@@ -63,6 +63,7 @@ class WXDLLIMPEXP_CORE wxTextEntry : public wxTextEntryBase
     bool GTKEntryOnInsertText(const char* text);
     bool GTKIsUpperCase() const { return m_isUpperCase; }
 
+    int GTKGetMaxLength() const { return m_maxlen; }
```
Nitpicking, but also here I don't understand your use of whitespace. Why do you group the function together with a totally unrelated comment for another function?

oneeyeman1 committed:
Implement PR suggested fixes

oneeyeman1 committed:
Indentation fix

oneeyeman1 commented:
@MaartenBent ,
Sorry for the delay in updating it - was trying to make wxFSVolume implemented on GTK/Linux.

Please update and re-test. I think I fixed all the issues mentioned.

TIA!!!

MaartenBent commented on src/gtk/textctrl.cpp:
```diff
@@ -796,9 +819,6 @@ bool wxTextCtrl::Create( wxWindow *parent,
         // work around probable bug in GTK+ 2.18 when calling WriteText on a
         // new, empty control, see https://github.com/wxWidgets/wxWidgets/issues/11409
         gtk_entry_get_text((GtkEntry*)m_text);
-
-        if (style & wxNO_BORDER)
-            gtk_entry_set_has_frame((GtkEntry*)m_text, FALSE);
```
Now the diff shows this as removed. I think only the `#ifndef __WXGTK3__` should have been removed, and the rest restored to how it originally was. Unless there is a reason why this should be removed, or not used with GTK3.

vadz replied:
This absolutely should be restored, no idea why was it removed.

MaartenBent replied:
This remark is still valid. The diff shows this code is still removed.

oneeyeman1 replied:
@MaartenBent ,
I don't know what diff you are looking at.
This code is restored on my local copy and in Git...

MaartenBent reviewed (commented):
Looks good enough now, just some small remarks. I think it can move on to the next reviewer.

But PLEASE PLEASE  PLEASE mark all the outdated and resolved comments as resolved, it makes reviewing so much easier.

vadz commented on src/gtk/textctrl.cpp:
```diff
@@ -1525,6 +1559,12 @@ void wxTextCtrl::GTKOnTextChanged()
 {
     if ( IgnoreTextUpdate() )
         return;
+    if( IsMultiLine() )
```
```suggestion

        if ( IsMultiLine() )
```

vadz commented on src/gtk/textctrl.cpp:
```diff
@@ -1525,6 +1559,12 @@ void wxTextCtrl::GTKOnTextChanged()
+    if( IsMultiLine() )
+    {
+        auto count = gtk_text_buffer_get_char_count( m_buffer );
+        if( m_maxlen > 0 && count >= m_maxlen )
```
Why is `>=` used here? Shouldn't it be `>`?

```suggestion
        if ( m_maxlen > 0 && count > m_maxlen )
```

oneeyeman1 replied:
@vadz , Changing this will bring back bad event generation behavior.

vadz replied:
I have no idea what is this supposed to mean, but it doesn't seem obvious at all why we should ignore the changes when the char count is exactly equal to `m_maxlen`, so you need to either explain it (in a comment) if it's somehow correct or fix the code if it isn't.

oneeyeman1 replied:
@vadz ,
I tried to explain it in the comment. Please let me know if you need a better wording...
Thx.

vadz replied:
This doesn't explain anything. When the count is equal to `m_maxlen`, there should be no events about overflowing it at all, so if anything, this indicates that there is a bug somewhere.

MaartenBent replied:
This seems to happens because after typing the character that exceeds the max, it is deleted in `au_insert_text_callback`.
But this delete action generates another gtk text event that calls this function.

I think the correct fix is to add this after `gtk_text_buffer_delete`:
```c++
win->IgnoreNextTextUpdate();
```

With this change, it seems all this code in `GTKOnTextChanged` is not needed anymore.
(But please test)

oneeyeman1 replied:
@MaartenBent ,
Made the changes and they seemed to work.

Could you try them please if you have a spare minute, please?

MaartenBent replied:
I tried the text example in Ubuntu gtk3 and there it works when typing and pasting.

oneeyeman1 replied:
@MaartenBent ,
Thx for testing.

Hopefully it is now good.

vadz commented on src/gtk/textctrl.cpp:
```diff
@@ -1227,6 +1257,10 @@ void wxTextCtrl::WriteText( const wxString &text )
 {
     wxCHECK_RET( m_text != nullptr, wxT("invalid text ctrl") );
 
+    auto old_maxlen = m_maxlen;
+    m_maxlen = 0;
+    wxON_BLOCK_EXIT_SET( m_maxlen, old_maxlen );
```
```suggestion
    // Disable max length checks, they don't apply to changes done by the program.
    auto maxlenOrig = m_maxlen;
    m_maxlen = 0;
    wxON_BLOCK_EXIT_SET( m_maxlen, maxlenOrig );
```

vadz commented on src/gtk/textctrl.cpp:
```diff
@@ -913,6 +933,16 @@ GtkEditable *wxTextCtrl::GetEditable() const
 
+void wxTextCtrl::SetMaxLength(unsigned long length)
+{
+    if( IsMultiLine() )
```
```suggestion
    if ( IsMultiLine() )
```

vadz commented on src/gtk/textctrl.cpp:
```diff
@@ -913,6 +933,16 @@ GtkEditable *wxTextCtrl::GetEditable() const
+        m_maxlen = length;
+    }
+    else
+        wxTextEntry::SetMaxLength( length );
```
```suggestion
    {
        wxTextEntry::SetMaxLength(length);
    }
```

vadz commented on src/gtk/textctrl.cpp:
```diff
@@ -560,6 +560,29 @@ au_insert_text_callback(GtkTextBuffer *buffer,
     }
+    auto maxlen = win->GTKGetMaxLength();
+    auto count = gtk_text_buffer_get_char_count( buffer );
+    if( maxlen > 0 && count > maxlen )
```
```suggestion
    if ( maxlen > 0 && count > maxlen )
```

vadz commented on src/gtk/textctrl.cpp:
```diff
@@ -560,6 +560,29 @@ au_insert_text_callback(GtkTextBuffer *buffer,
         wxGtkTextApplyTagsFromAttr(win->GetHandle(), buffer, win->GetDefaultStyle(),
                                    &start, end);
     }
+    auto maxlen = win->GTKGetMaxLength();
```
```suggestion
    const auto maxlen = win->GTKGetMaxLength();
```

vadz commented on src/gtk/textctrl.cpp:
```diff
@@ -560,6 +560,29 @@ au_insert_text_callback(GtkTextBuffer *buffer,
+        // reason, ("Run Last") it won't work in GTKOnInsertText() as it called from
+        // the handler that does not connected after
+        gtk_text_buffer_delete( buffer, &offset, end );
+#if GTK_CHECK_VERSION( 3, 2, 0 )
+        if( wx_is_at_least_gtk3(2) )
+            gtk_text_iter_assign( end, &offset );
+#endif
```
What is this code for? `end` doesn't seem to be used any more, so why do we bother changing it? And, of course, if we do need to change it, what's wrong with just writing `end = offset`??

```suggestion
```

oneeyeman1 replied:
@vadz , I tried to get rid of this and didn't see any bad side effects. Lets's see how it will work with other GTK+ version...

vadz reviewed (changes requested):
Thanks a lot to Maarten for taking time to go through all this, but there are still things to fix.

@oneeyeman1 I **very strongly** recommend rereading the diff of your changes after submitting it.

oneeyeman1 committed:
Implement review suggestions

oneeyeman1 committed:
Implement the last suggestion

vadz commented on src/gtk/textctrl.cpp:
```diff
@@ -560,6 +560,25 @@ au_insert_text_callback(GtkTextBuffer *buffer,
                                    &start, end);
     }
+    const auto maxlen = win->GTKGetMaxLength();
+    auto count = gtk_text_buffer_get_char_count( buffer );
+    if ( maxlen > 0 && count > maxlen )
```
It's pointless to call `gtk_text_buffer_get_char_count()` if `maxlen` is not set, so this should be:

```suggestion
    if ( maxlen > 0 )
    {
        const auto count = gtk_text_buffer_get_char_count( buffer );
        if ( count > maxlen )
```

(maybe this function is cheap to call, but maybe it isn't, and it's simpler to not to even have to think about it by avoiding calling it unless necessary).

oneeyeman1 committed:
Code review suggestion

oneeyeman1 committed:
Add an explanation comment

oneeyeman1 committed:
Simlify the code according to code review suggestion

vadz commented:
This needs to be rebased/force pushed to resolve the conflicts.

oneeyeman1 commented:
@vadz ,

I tried to rebase but I need some guidance as I think something is not right.

1. I ran "git remote add upstream https://github.com/wxWidgets/wxWidgets" amd confirm with "git remote -v" that it is displayed:

origin	git@github.com:oneeyeman1/wxWidgets.git (fetch)
origin	git@github.com:oneeyeman1/wxWidgets.git (push)
upstream	https://github.com/wxWidgets/wxWidgets (fetch)
upstream	https://github.com/wxWidgets/wxWidgets (push)

2. I ran "git fetch upstream"
3. Ran "git rebase upstream/master"
4. Had one conflict. Fixed it and then ran "git add interface/wx/textentry.h" and then "git rebase --continue"
5. Got the message "Successfully rebased and updated refs/heads/MultilineGTKMaxLength_v2"
6. Next did:

igor@WaylandGnome ~/wx_copy $ git status
On branch MultilineGTKMaxLength_v2
Your branch and 'origin/MultilineGTKMaxLength_v2' have diverged,
and have 753 and 16 different commits each, respectively.
  (use "git pull" if you want to integrate the remote branch with yours)

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   src/expat (new commits)
	modified:   src/stc/lexilla (new commits)

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	buildGTK/
	patch

no changes added to commit (use "git add" and/or "git commit -a")

What am I suppose to do now?

The conflicted file is not there to add/commit. I can add expat/lexilla,

Thx for the help...

vadz commented:
Force push (`git push --force`) your branch here.

MaartenBent commented:
Yes, just force push after the successful rebase.

Definitely NOT add expat/lexilla. You can (optionally) get those expat/lexilla submodules up-to-date by running `git submodule update`.

oneeyeman1 commented:
Thx, guys.

I did force-push.

Lets wait and see what happens...

oneeyeman1 commented:
@vadz ,
Can you also apply this one?

Thx.
